src/CoinEntry.h

```diff
@@ -0,0 +1,55 @@
+// Copyright © 2017-2020 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#pragma once
+
+#include <TrustWalletCore/TWCoinType.h>
+
+#include "Data.h"
+#include "PublicKey.h"
+
+#include <string>
+#include <vector>
+
+namespace TW {
+
+/// Interface for coin-specific entry, used to dispatch calls to coins
+/// Implement this for all coins.
+class CoinEntry {
```

Comment:
I think it's the right direction to move, I'd like to also have `Address` and `Signer` interface
For `Address`, it's basically the same methods with `AnyAddress`
1. create with String
2. create with public key (hide the p2pkh, p2sh prefix, hrp implemention details, to its own file)

For `Signer`: 
```c++
static Proto::SigningOutput sign(const Proto::SigningInput& input) noexcept;
```

We also codegen configuration from `coins.json`, maybe we can replace that part as well, even more `switch(coin)` in `src/Generated/Coins.cpp`